raster/src/main/scala/geotrellis/raster/GridExtent.scala

```diff
@@ -2,29 +2,115 @@ package geotrellis.raster
+    val xmax = extent.xmax - (math.floor((extent.xmax - targetExtent.xmax) / cellwidth) * cellwidth)
+    val ymin = extent.ymin + (math.floor((targetExtent.ymin - extent.ymin) / cellheight) * cellheight)
+    val ymax = extent.ymax - (math.floor((extent.ymax - targetExtent.ymax) / cellheight) * cellheight)
+
+    GridExtent(Extent(xmin, ymin, xmax, ymax), cellwidth, cellheight)
+  }
+
+  /**
+    * Tests if the grid is aligned to the extent.
+    * This is true when the extent is evenly divided by cellheight and cellwidth.
+    */
+  def isGridExtentAlligned(): Boolean = {
+    def isWhole(x: Double) = math.floor(x) == x
+    isWhole((extent.xmax - extent.xmin) / cellwidth) && isWhole((extent.ymax - extent.ymin) / cellheight)
+  }
 
   /**
-    * An implicit function to convert a [[GridDefinition]] to a
-    * [[GridExtent]].
+    * Returns a RasterExtent that lines up with this RasterExtent's
+    * resolution, and grid layout.
+    *
+    * For example, the resulting RasterExtent will not have the given
+    * extent, but will have the smallest extent such that the whole of
+    * the given extent is covered, that lines up with the grid.
     */
-  implicit def gridDefinitionToGridExtent(gridDefinition: GridDefinition): GridExtent =
-    gridDefinition.toGridExtent
+  def createAlignedRasterExtent(targetExtent: Extent): RasterExtent =
+    createAlignedGridExtent(targetExtent).toRasterExtent
+
+  /**
+    * Gets the Extent that matches the grid bounds passed in, aligned
+    * with this RasterExtent.
+    *
+    * The 'clamp' parameter determines whether or not to clamp the
+    * Extent to the extent of this RasterExtent; defaults to true. If
+    * true, the returned extent will be contained by this
+    * RasterExtent's extent, if false, the Extent returned can be
+    * outside of this RasterExtent's extent.
+    *
+    * @param  gridBounds  The extent to get the grid bounds for
+    * @param  clamp       A boolean which controlls the clamping behvior
+    */
+  def extentFor(gridBounds: GridBounds, clamp: Boolean = true): Extent = {
+    val xmin = gridBounds.colMin * cellwidth + extent.xmin
+    val ymax = extent.ymax - (gridBounds.rowMin * cellheight)
+    val xmax = xmin + (gridBounds.width * cellwidth)
+    val ymin = ymax - (gridBounds.height * cellheight)
+
+    if(clamp) {
+      Extent(
+        max(min(xmin, extent.xmax), extent.xmin),
+        max(min(ymin, extent.ymax), extent.ymin),
+        max(min(xmax, extent.xmax), extent.xmin),
+        max(min(ymax, extent.ymax), extent.ymin)
+      )
+    } else {
+      Extent(xmin, ymin, xmax, ymax)
+    }
+  }
+}
+
+
+object GridExtent {
+  def apply(extent: Extent, cellSize: CellSize): GridExtent =
+    new GridExtent(extent, cellSize.width, cellSize.height)
+
+  def apply(extent: Extent, cellwidth: Double, cellheight: Double): GridExtent =
+    new GridExtent(extent, cellwidth, cellheight)
```

Comment:
We lose an unapply here by taking away it's case classy-ness. Also proper equality. I get that you want no case class for inheritance - we should provide that functionality ourselves.